# Trabajo Práctico 3: Predicción de tormenta en base de Datos Meteorológicos usando modelo más avanzados

### Alumno: Luis Gabriel Rueda Yarlequé

### 1. Modelado y Evaluación

Se selecciona el modelo de bosques aleatorios

In [1]:
# Importamos librerías necesarias
import pandas as pd
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import balanced_accuracy_score, f1_score

In [2]:
# Cargamos los datos

df = pd.read_csv('weatherAUS_limpio.csv')
x_test = pd.read_csv('x_test.csv')
y_test = pd.read_csv('y_test.csv')
x_train = pd.read_csv('x_train.csv')
y_train = pd.read_csv('y_train.csv')

In [3]:
# Implementación del modelo

random_forest_classifier = RandomForestClassifier(random_state = 42)
random_forest_classifier.fit(x_train, y_train)

/home/broklk/.pyenv/versions/3.10.14/envs/venv-3.10/lib/python3.10/site-packages/sklearn/base.py:1473: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return fit_method(estimator, *args, **kwargs)


RandomForestClassifier(random_state=42)

In [4]:
# Realizamos predicciones

y_pred = random_forest_classifier.predict(x_test)

#### 1.1. Evalúe el modelo con las métricas definidas.

In [5]:
# Calculamos métricas de evaluación (hiperparámetros por defecto)

balanced_accuracy = balanced_accuracy_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)

print(f"Precisión Balanceada: {balanced_accuracy*100:.2f}%")
print(f"F1-score: {f1*100:.2f}%")

Precisión Balanceada: 72.65%
F1-score: 60.26%


#### 1.2. Discuta las fortalezas y debilidades del modelo elegido en el contexto específico de este problema.

##### Fortalezas:

- A diferencia de un solo árbol de decisión, los bosques aleatorios son menos propensos a sobreajustes a los datos de entrenamiento debido a la agregación de resultados de multiples árboles. Por lo cual, los bosques aleatorios son mas robustos en la predicción de lluvia que un solo árbol.
- Los bosques aleatorios son buenos para relaciones complejas entre distintas variables, como la humedad, temperatura y presión.
- Los bosques aleatorios pueden manejar conjuntos de datos grandes y de alta dimensionalidad, lo que es beneficioso si se dispone de datos meteorológicos variados.

##### Debilidades:

- Los árboles de decisión individuales son fáciles de interpretar, los bosques aleatorios, que consisten en cientos o miles de árboles, son mucho más difíciles de interpretar.
- Entrenar un modelo de Random Forest suele ser más costoso en términos de tiempo y recursos computacionales, especialmente cuando se trabaja con un gran número de árboles y un gran conjunto de datos extenso.
- Aunque los bosques aleatorios son robustos, el rendimiento puede depender de una buena selección de hiperparámetros. Si estos parámetros no se ajustan correctamente, el modelo puede no funcionar de manera óptima.


### 2. Conclusiones

#### 2.1. Construya una tabla comparativa con los resultados de los diferentes modelos y el modelo base. ¿Cuál fue el mejor modelo y por qué?

| Modelo              | Precisión balanceada | F1     |
|---------------------|----------------------|--------|
| Base                | 65.47%               | 46%    |
| KNN                 | 64.21%               | 44.10% |
| Regresión Logística | 72.52%               | 59.46% |
| Árboles aleatorios  | 72.65%               | 60.26% |

El mejor modelo es el de **Árboles aleatorios**, porque presenta mejores métricas de evaluación y nos va a permitir tener una mejor predicción de las lluvias.

#### 2.2. Resuma los hallazgos más relevantes obtenidos durante todo el proceso desde el TP1 al TP3.

##### **TP1**

- Se logró realizar el análisis del dataset del clima de Australia y realizar la limpieza de datos, de aquellos atributos que estaban en *Null*, la importancia de estos primeros pasos son cruciales para los análisis posteriores.
- Otro punto importante es manejar el formato correcto en cada parámetro.
- Se puede realizar análisis por medio de medidas estadísticas de los atributos, asi como también, a traves de gráficas que permiten evidenciar patrones, estacionalidad, correlación, etc.

##### **TP2**

- Se vió la importancia del un preprocesamiento de datos antes del entrenamiento de un dataframe, también la estandarización y/o normalización de los valores numéricos.
- Siempre se debe realizar una separación de datos de entrenamiento y de pruebas.
- La decisión de que medidas de evaluación se usarán para evaluar el performance de los modelos es de suma importancia, ya que su elección también debe darse de acuerdo al contexto del problema y los datos que se evalúan.
- Los modelos más básicos también pueden ser mejores que algunos métodos más elaborados, en ciertas circunstancias.

##### **TP3**

- El probar distintos hiperparámetros en los modelos es importantes para la implementación de los mismos, sin embargo, también significan un coste computacional que no siempre puede ser asumido.
- Entre mas precisión se busca en un modelo, se suele tener mayor coste computacional en el entrenamiento del modelo.
- Modelos como *Árboles aleatorios* suelen ser muy robustos y con una buena elección de parámetros puede ser muy preciso.

#### 2.3. Guarde en formato pickle el mejor modelo, seleccionado según su criterio.

In [6]:
import pickle

with open('random_forest_classifier.pkl', 'wb') as f:
    pickle.dump(random_forest_classifier, f)

print("El modelo ha sido guardado satisfactoriamente")

El modelo ha sido guardado satisfactoriamente
